sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1939,11 +1939,32 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
     Literal(BigDecimal(ctx.getText).underlying())
   }
 
+  /**
+   * Create a decimal literal for a regular decimal number.
+   */
+  override def visitLegacyDecimalLiteral(ctx: LegacyDecimalLiteralContext)
+    : Literal = withOrigin(ctx) {
+    Literal(BigDecimal(ctx.getText).underlying())
+  }
+
+  /**
+   * Create a double literal for number with an exponent, e.g. 1E-30
+   */
+  override def visitExponentLiteral(ctx: ExponentLiteralContext): Literal = {
+    numericLiteral(ctx, Double.MinValue, Double.MaxValue, DoubleType.simpleString)(_.toDouble)
```

Comment:
What `numericLiteral` does is to strip the suffix (which we don't have in expenent) and check range (is it really useful for double?)

Reply:
For exponent values, we recognize it by searching "E" in `numericLiteral`, see:

```
val rawStrippedQualifier = if (text.toUpperCase(Locale.ROOT).contains("E")) {
  // exponent values don't have a suffix but contain 'E'
  text
} else {
  text.substring(0, ctx.getText.length - 1)
}
```

Reply:
hmm this looks hacky. Can we move the strip of suffix to the caller side? then `numericLiteral` can focus on the range check.

Reply:
Sounds good.